server/odc-service/src/main/java/com/oceanbase/odc/service/monitor/InMemoryMetricManager.java

```diff
@@ -0,0 +1,176 @@
+
+    private final Map<MeterKey, CounterWrapper> COUNTER_MAP = new ConcurrentHashMap<>();
+    private final Map<Pair<String, MeterKey>, TimerSampleHolder> TIMER_SAMPLE_MAP = new ConcurrentHashMap<>();
+    @Autowired
+    private BusinessMeterRegistry businessMeterRegistry;
+    @Autowired
+    private MonitorProperties monitorProperties;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        registerGauge(MeterKey.ofMeter(MeterName.METER_COUNTER_HOLDER_COUNT), this.COUNTER_MAP::size);
+        registerGauge(MeterKey.ofMeter(MeterName.METER_TIMER_HOLDER_COUNT), this.TIMER_SAMPLE_MAP::size);
+    }
+
+
+    public void registerGauge(MeterKey meterKey, Supplier<Number> f) {
+        MeterName meterName = meterKey.getMeterName();
+        Iterable<Tag> tags = meterKey.getTags();
+        Gauge.Builder<Supplier<Number>> builder = Gauge.builder(meterName.getMeterName(), f)
+                .description(MeterName.CONNECT_SESSION_ACTIVE_COUNT.getDescription());
+        for (Tag tag : tags) {
+            builder.tag(tag.getKey(), tag.getValue());
+        }
+        builder.register(businessMeterRegistry);
+    }
+
+    public void incrementCounter(MeterKey meterKey) {
+        CounterWrapper wrapper = COUNTER_MAP.get(meterKey);
+        if (wrapper == null) {
+            if (COUNTER_MAP.size() >= monitorProperties.getMeter().getMaxCounterMeterNumber()) {
+                log.error("Too many counter to register, meterKey={}", meterKey);
+                return;
+            }
+            COUNTER_MAP.computeIfAbsent(meterKey, (key) -> new CounterWrapper(null));
+        }
+        MeterKey lockKey = getCounterLockKey(meterKey);
+        synchronized (lockKey) {
+            wrapper = COUNTER_MAP.get(lockKey);
+            if (wrapper.getCounter() == null) {
+                Counter counter = registerCounter(meterKey);
+                wrapper.setCounter(counter);
+            }
+            wrapper.getCounter().increment();
+        }
+    }
+
+    private MeterKey getCounterLockKey(MeterKey meterKey) {
+        return COUNTER_MAP.keySet().stream().filter(meterKey::equals).findFirst()
+                .orElseThrow(IllegalAccessError::new);
+    }
+
+    public Counter registerCounter(MeterKey meterKey) {
+        Builder builder = Counter.builder(meterKey.getMeterName().getMeterName())
+                .description(meterKey.getMeterName().getDescription());
+        for (Tag tag : meterKey.getTags()) {
+            builder.tag(tag.getKey(), tag.getValue());
+        }
+        return builder.register(businessMeterRegistry);
+    }
+
+    public void startTimerSample(String sampleKey, MeterKey meterKey) {
+        Pair<String, MeterKey> sampleMapKey = Pair.of(sampleKey, meterKey);
+        TimerSampleHolder sampleHolder = this.TIMER_SAMPLE_MAP.get(sampleMapKey);
+        if (sampleHolder == null) {
+            if (this.TIMER_SAMPLE_MAP.size() >= monitorProperties.getMeter().getMaxTimerMeterNumber()) {
```

Comment:
As a code designer, if you are afraid of memory leak to add protection code instead of design a better mechanism to prevent memory leak, then the code is hard to maintain
As a user, I thought i have start timer sample, but nothing done, Is that what you expected?

Reply:
Setting a maximum allowed value is a common practice. For instance, WebMvcMetrics operates similarly, the only difference being the strategy for handling values that exceed the maximum, where, apart from logging, alarm are also added.